components/eam/src/physics/cam/hetfrz_classnuc.F90

```diff
@@ -761,128 +770,236 @@ pure function get_Aimm(vwice, rgimm, temperature, dg0imm) result(Aimm)
+   real(r8), intent(in) :: Tdiff            ! temperature difference between droplet and environment [K]
+ 
+   ! output
+   real(r8), intent(out) :: K_total         ! collision kernel [cm^3/s]
+
+   ! local variables
+   real(r8) :: Kn          ! Knudsen number [unitless]
+   real(r8) :: Sc          ! Schmidt number [unitless]
+   real(r8) :: Daer        ! aerosol diffusivity [m2/s] 
+   real(r8) :: f_t         ! factor by Waldmann & Schmidt [unitless]
+   real(r8) :: Q_heat      ! heat flux [J/m2/s]
+   real(r8) :: K_brownian,K_thermo_cotton,K_diffusio_cotton   ! collision kernels [m3/s]
+
+   ! Note for C++ port: Due to BFB for Fortran code, we have to declare
+   ! Boltzmann constant here again. This value is only used in this
+   ! subroutine.
+   real(r8), parameter :: kboltz2 = 1.38065e-23_r8      ! Boltzmann constant [J/K]
+
+         
+   ! Knudsen number (Seinfeld & Pandis 8.1)
+   Kn = lambda/r_a
+      
+   ! aerosol diffusivity
+   Daer = kboltz2*temperature*(1 + Kn)/(6*pi*r_a*viscos_air)
+      
+   ! Schmidt number
+   Sc = viscos_air/(Daer*rho_air)
+
+   ! Young (1974) first equ. on page 771
+   K_brownian = 4*pi*r3lx*Daer*(1 + 0.3_r8*Re**0.5_r8*Sc**0.33_r8)
+
+   ! form factor
+   f_t = 0.4_r8*(1._r8 + 1.45_r8*Kn + 0.4_r8*Kn*EXP(-1._r8/Kn))      &
                   *(Ktherm_air + 2.5_r8*Kn*Ktherm)                      &
                /((1._r8 + 3._r8*Kn)*(2._r8*Ktherm_air + 5._r8*Kn*Ktherm+Ktherm))
-      ! calculate T-Tc as in Cotton et al.
-      Tdiff_cotton = -G*(rhwincloud - 1._r8)*latvap/Ktherm_air
-      Q_heat = Ktherm_air/r3lx*(1._r8 + 0.3_r8*Re**0.5_r8*Pr**0.33_r8)*Tdiff_cotton
-      K_thermo_cotton = 4._r8*pi*r3lx*r3lx*f_t*Q_heat/pres
-      K_diffusio_cotton = -(1._r8/f_t)*(rh2o*t/latvap)*K_thermo_cotton
-      K_total = 1.e6_r8*(K_brownian + K_thermo_cotton + K_diffusio_cotton)  ! convert m3/s -> cm3/s
-      ! set K to 0 if negative
-      if (K_total .lt. 0._r8) K_total = 0._r8
-
-      if (i == 1) Kcoll_bc = K_total
-      if (i == 2) Kcoll_dust_a1 = K_total
-      if (i == 3) Kcoll_dust_a3 = K_total
-        
-   end do
-      
-end subroutine collkernel
+
+   Q_heat = Ktherm_air/r3lx*(1._r8 + 0.3_r8*Re**0.5_r8*Pr**0.33_r8)*Tdiff
+
+   K_thermo_cotton = 4._r8*pi*r3lx*r3lx*f_t*Q_heat/pressure
+
+   K_diffusio_cotton = -(1._r8/f_t)*(rh2o*temperature/latvap)*K_thermo_cotton
+   K_total = 1.e6_r8*(K_brownian + K_thermo_cotton + K_diffusio_cotton)  ! convert m3/s -> cm3/s
+
+   ! set K to 0 if negative
+   if (K_total < 0._r8) K_total = 0._r8
+
+end subroutine
+
+
+pure function get_air_viscosity(tc) result(viscos_air)
+
+   implicit none
+   real(r8), intent(in) :: tc  ! temperature [deg C]
+
+   real(r8) :: viscos_air      ! dynamic viscosity of air [kg/m/s]
+   real(r8), parameter :: coeff_a = -1.2e-5_r8
+   real(r8), parameter :: coeff_b = 0.0049_r8
+   real(r8), parameter :: coeff_c = 1.718_r8
+
+   viscos_air = (coeff_c + coeff_b*tc + coeff_a*tc*tc)*1.e-5_r8
+
```

Comment:
NBFB issue caused by this line, fixed by changing calculation sequence and make tc**2 to *tc*tc